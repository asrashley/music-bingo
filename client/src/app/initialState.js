import { initialState as cards } from '../cards/cardsSlice';
import { initialState as games } from '../games/gamesSlice';
import { initialState as tickets } from '../tickets/ticketsSlice';
import { initialState as user } from '../user/userSlice';

export const initialState = Object.freeze({
    cards,
    games,
    tickets,
    user,
});
