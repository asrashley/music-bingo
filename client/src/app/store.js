import { configureStore, getDefaultMiddleware } from '@reduxjs/toolkit';
import logger from 'redux-logger';
import { connectRouter } from 'connected-react-router';
import { createBrowserHistory } from 'history';
import { reducer as formReducer } from 'redux-form';

import cardsReducer from '../cards/cardsSlice';
import gamesReducer from '../games/gamesSlice';
import ticketsReducer from '../tickets/ticketsSlice';
import userReducer from '../user/userSlice';
import adminReducer from '../admin/adminSlice';
import usersMiddleware from '../user/userMiddleware';

export const history = createBrowserHistory();

export const store = configureStore({
  middleware: [...getDefaultMiddleware(), usersMiddleware, logger],
  reducer: {
    admin: adminReducer,
    cards: cardsReducer,
    form: formReducer,
    games: gamesReducer,
    tickets: ticketsReducer,
    router: connectRouter(history),
    user: userReducer,
  },
});
