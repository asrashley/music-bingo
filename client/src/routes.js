
export default {
  index: '/',
  login: '/user/login',
  logout: '/user/logout',
  passwordReset: '/user/reset',
  passwordResetConfirm: '/user/reset/:token',
  register: '/register',
  game: '/game/:gamePk',
  play: '/game/:gamePk/tickets',
  viewTicket: '/game/:gamePk/:ticketPk',
  pastGames: '/history',
  trackListing: '/history/:gamePk',
  listUsers: '/users'
};

