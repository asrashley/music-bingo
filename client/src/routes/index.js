export default {
  index: '/',
  login: '/user/login',
  logout: '/user/logout',
  user: '/user',
  changeUser: '/user/modify',
  passwordReset: '/user/reset',
  passwordResetConfirm: '/user/reset/:token',
  guestLinks: '/user/guests',
  listUsers: '/user/users',
  guestAccess: '/invite/:token',
  register: '/register',
  listDirectories: '/directories',
  listGames: '/game',
  chooseTickets: '/game/:gameId',
  play: '/game/:gameId/tickets',
  viewTicket: '/game/:gameId/:ticketPk',
  pastGames: '/history',
  trackListing: '/history/:gameId',
  privacy: '/privacy',
};
