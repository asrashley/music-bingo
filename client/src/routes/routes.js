

export const routes = Object.freeze({
  index: '/',
  login: '/user/login',
  logout: '/user/logout',
  user: '/user',
  changeUser: '/user/modify',
  passwordReset: '/user/reset',
  passwordResetConfirm: '/user/reset/:token',
  guestLinks: '/user/guests',
  settingsIndex: '/user/settings',
  settingsSection: '/user/settings/:section',
  listUsers: '/user/users',
  guestAccess: '/invite/:token',
  register: '/register',
  listDirectories: '/clips',
  listDirectory: '/clips/:dirPk',
  listGames: '/game',
  chooseTickets: '/game/:gameId',
  play: '/game/:gameId/tickets',
  viewTicket: '/game/:gameId/tickets/:ticketPk',
  pastGamesPopularity: '/history',
  pastGamesList: '/history/games',
  pastGamesCalendar: '/history/calendar',
  pastGamesCalendarByYear: '/history/calendar/:year',
  pastGamesByTheme: '/history/themes/:slug',
  gameLastUsed: '/history/themes',
  trackListingByTheme: '/history/themes/:slug/:gameId',
  trackListing: '/history/games/:gameId',
  privacy: '/privacy',
});
