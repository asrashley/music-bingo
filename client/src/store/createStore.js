import { configureStore, getDefaultMiddleware } from '@reduxjs/toolkit';
import logger from 'redux-logger';
import { connectRouter } from 'connected-react-router';

import adminReducer from '../admin/adminSlice';
import directoriesReducer from '../directories/directoriesSlice';
import gamesReducer from '../games/gamesSlice';
import messagesReducer from '../messages/messagesSlice';
import settingsReducer from '../settings/settingsSlice';
import systemReducer from '../system/systemSlice';
import ticketsReducer from '../tickets/ticketsSlice';
import userReducer from '../user/userSlice';
import usersMiddleware from '../user/userMiddleware';
import messagesMiddleware from '../messages/messagesMiddleware';

import { history } from './history';

export function createStore(preloadedState, logging) {
  const middleware = [...getDefaultMiddleware(), usersMiddleware, messagesMiddleware];
  if (logging === undefined) {
    logging = (process.env.NODE_ENV === 'development');
  }
  if (logging) {
    middleware.push(logger);
  }
  return configureStore({
    middleware,
    reducer: {
      admin: adminReducer,
      directories: directoriesReducer,
      games: gamesReducer,
      messages: messagesReducer,
      settings: settingsReducer,
      system: systemReducer,
      tickets: ticketsReducer,
      router: connectRouter(history),
      user: userReducer,
    },
    preloadedState
  });
}
