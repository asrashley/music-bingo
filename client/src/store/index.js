export { createStore } from './createStore';
export { initialState } from './initialState';
export { history } from './history';
