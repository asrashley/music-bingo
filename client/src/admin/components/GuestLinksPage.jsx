import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { reverse } from 'named-urls';
import { Link } from 'react-router-dom';

import { DateTime } from '../../components';

/* actions */
import { fetchUserIfNeeded } from '../../user/userSlice';
import {
  fetchGuestLinksIfNeeded, createGuestToken, deleteGuestToken,
  invalidateGuestTokens
} from '../adminSlice';

/* selectors */
import { getUser } from '../../user/userSelectors';
import { getGuestTokens, getAdminUserPk } from '../adminSelectors';

/* data */
import { initialState } from '../../app/initialState';
import routes from '../../routes';

function TableRow({token, onDelete}) {
  const link = reverse(`${routes.guestAccess}`, { token:token.jti});
  return (
    <tr>
      <td className="link">
        <Link to={link}>{link}</Link>
      </td>
      <td className="expires"><DateTime date={token.expires}/></td>
      <td className="delete">
        <button className="btn btn-warning"
        onClick={() => onDelete(token)}>Delete</button>
      </td>
    </tr>
  );
}

class GuestLinksPage extends React.Component {
  static propTypes = {
    dispatch: PropTypes.func.isRequired,
    history: PropTypes.object.isRequired,
    user: PropTypes.object.isRequired,
  };

  componentDidMount() {
    const { dispatch, user } = this.props;
    dispatch(fetchUserIfNeeded());
    if(user.loggedIn) {
      dispatch(fetchGuestLinksIfNeeded());
    }
  }

  componentDidUpdate(prevProps, prevState) {
    const { dispatch, user, userPk } = this.props;
    if (userPk !== prevProps.userPk && user.loggedIn) {
      console.log(`user.pk=${user.pk}`);
      dispatch(fetchGuestLinksIfNeeded());
    }
  }

  addLink = () => {
    const { dispatch } = this.props;
    dispatch(createGuestToken());
  }

  deleteLink = (token) => {
    const { dispatch } = this.props;
    dispatch(deleteGuestToken(token.jti));
  }

  refreshTokenList = () => {
    const { dispatch } = this.props;
    dispatch(invalidateGuestTokens());
    dispatch(fetchGuestLinksIfNeeded());
  }

  render() {
    const { tokens } = this.props;
    return(
      <div className="guest-tokens">
        <table className="table table-striped table-bordered">
          <thead>
            <tr>
              <th className="link">Link</th>
              <th className="expires">Expires</th>
              <th className="delete">
                Delete
                <button className="btn btn-light refresh-icon btn-sm float-right"
                        onClick={this.refreshTokenList}>&#x21bb;</button>
              </th>
            </tr>
          </thead>
          <tbody>
            {tokens.map((token, idx) => <TableRow key={idx} token={token} onDelete={this.deleteLink}/>)}
          </tbody>
          <tfoot>
            <tr>
              <td colSpan="3">
                <button className="btn btn-primary" onClick={this.addLink}>Add link</button>
              </td>
            </tr>
          </tfoot>
        </table>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
    state = state || initialState;
    return {
      user: getUser(state, props),
      tokens: getGuestTokens(state, props),
      userPk: getAdminUserPk(state, props),
    };
  };

 GuestLinksPage = connect(mapStateToProps)(GuestLinksPage);

  export { GuestLinksPage };
