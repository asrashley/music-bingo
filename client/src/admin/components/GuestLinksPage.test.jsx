import React from 'react';
import { act, fireEvent, getByRole, waitFor } from '@testing-library/react';
import log from 'loglevel';
import { reverse } from 'named-urls';
import fetchMock from 'fetch-mock';

import { renderWithProviders, installFetchMocks } from '../../../tests';
import { createStore } from '../../store/createStore';
import { initialState } from '../../store/initialState';
import { routes } from '../../routes';
import { GuestLinksPage } from './GuestLinksPage';

import { adminUser } from '../../../tests/MockServer';

describe('GuestLinksPage component', () => {
  let apiMock = null;

  beforeAll(() => {
    vi.useFakeTimers();
  });

  beforeEach(() => {
    vi.setSystemTime(new Date('08 Apr 2023 18:02:00 GMT').getTime());
    apiMock = installFetchMocks(fetchMock, { loggedIn: true });
  });

  afterEach(() => {
    fetchMock.reset();
    log.resetLevel();
    vi.restoreAllMocks();
    vi.clearAllTimers();
    apiMock = null;
  });

  afterAll(() => {
    vi.useRealTimers();
  });

  it('to shows a list of guest links', async () => {
    const user = apiMock.getUserState(adminUser);
    const store = createStore({
      ...initialState,
      admin: {
        ...initialState.admin,
        user: user.pk,
        guest: {
          tokens: apiMock.guestTokens,
          invalid: false,
          error: null,
          lastUpdated: Date.now(),
        },
      },
      user
    });
    const { asFragment, findByText } = renderWithProviders(<GuestLinksPage />, { store });
    await Promise.all(apiMock.guestTokens.map((token) => {
      const link = reverse(`${routes.guestAccess}`, { token: token.jti });
      return findByText(link);
    }));
    expect(asFragment()).toMatchSnapshot();
  });

  it('reloads guest links', async () => {
    const user = apiMock.getUserState(adminUser);
    const store = createStore({
      ...initialState,
      admin: {
        ...initialState.admin,
        user: user.pk
      },
      user
    });
    const { findByText, findByLastUpdate } = renderWithProviders(<GuestLinksPage />, { store });
    const link = reverse(`${routes.guestAccess}`, { token: apiMock.guestTokens[0].jti });
    await findByText(link);
    const update = parseInt(document.querySelector('div.guest-tokens').dataset.lastUpdate, 10);
    act(() => {
      vi.advanceTimersByTime(150);
      vi.runOnlyPendingTimers();
    });
    fireEvent.click(document.querySelector('.refresh-icon'));
    await findByLastUpdate(update, { comparison: 'greaterThan' });
    expect(fetchMock.calls('/api/user/guest', 'GET').length).toEqual(2);
  });

  it('to allows a guest link to be added', async () => {
    const user = apiMock.getUserState(adminUser);
    const store = createStore({
      ...initialState,
      admin: {
        ...initialState.admin,
        user: user.pk
      },
      user
    });
    const addTokenPromise = apiMock.addResponsePromise('/api/user/guest/add', 'PUT');
    const { findByText, getByText } = renderWithProviders(<GuestLinksPage />, { store });
    const before = apiMock.guestTokens.length;
    fireEvent.click(getByText('Add link'));
    await waitFor(async () => {
      await addTokenPromise;
    })
    await fetchMock.flush(true);
    expect(fetchMock.calls('/api/user/guest/add', 'PUT').length).toEqual(1);
    expect(apiMock.guestTokens.length).toEqual(before + 1);
    const token = apiMock.guestTokens[apiMock.guestTokens.length - 1];
    await findByText(`/invite/${token.jti}`, { exact: false });
  });

  it('to allows a guest link to be deleted', async () => {
    const user = apiMock.getUserState(adminUser);
    const tokens = apiMock.guestTokens;
    const store = createStore({
      ...initialState,
      admin: {
        ...initialState.admin,
        user: user.pk,
        guest: {
          tokens,
          invalid: false,
          error: null,
          lastUpdated: Date.now(),
        },
      },
      user
    });
    const deleteLinkProm = apiMock.addResponsePromise(
      `/api/user/guest/delete/${tokens[0].jti}`, 'DELETE');
    const { findByText, getByTestId } = renderWithProviders(<GuestLinksPage />, { store });
    const link = reverse(`${routes.guestAccess}`, { token: tokens[0].jti });
    await findByText(link);
    const row = getByTestId(`token.${tokens[0].pk}`);
    fireEvent.click(getByRole(row, 'button'));
    await deleteLinkProm;
  });

});
