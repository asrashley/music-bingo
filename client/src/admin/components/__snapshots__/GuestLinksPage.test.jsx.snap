// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GuestLinksPage component to shows a list of guest links 1`] = `
<DocumentFragment>
  <div
    class="guest-tokens"
    data-last-update="1680976920000"
  >
    <table
      class="table table-striped table-bordered"
    >
      <thead>
        <tr>
          <th
            class="link"
          >
            Link
          </th>
          <th
            class="expires"
          >
            Expires
          </th>
          <th
            class="delete"
          >
            Delete
            <button
              class="btn btn-light refresh-icon btn-sm float-right"
            >
              â†»
            </button>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          data-testid="token.45"
        >
          <td
            class="link"
          >
            <a
              href="/invite/RyxeM-idLQ"
            >
              /invite/RyxeM-idLQ
            </a>
          </td>
          <td
            class="expires"
          >
            08:12:52 27/02/2023
          </td>
          <td
            class="delete"
          >
            <button
              class="btn btn-warning"
            >
              Delete
            </button>
          </td>
        </tr>
        <tr
          data-testid="token.46"
        >
          <td
            class="link"
          >
            <a
              href="/invite/cT34TCj7vg"
            >
              /invite/cT34TCj7vg
            </a>
          </td>
          <td
            class="expires"
          >
            08:12:54 27/02/2023
          </td>
          <td
            class="delete"
          >
            <button
              class="btn btn-warning"
            >
              Delete
            </button>
          </td>
        </tr>
      </tbody>
      <tfoot>
        <tr>
          <td
            colspan="3"
          >
            <button
              class="btn btn-primary"
            >
              Add link
            </button>
          </td>
        </tr>
      </tfoot>
    </table>
  </div>
</DocumentFragment>
`;
